### PR TITLE
Accept `nan` in `trial.report`

### PR DESCRIPTION
**Pull request by belldandyxtq**
## Motivation
This is a part of https://github.com/optuna/optuna/issues/3205 to support `nan` in `trial.report`. As shown in the issue, `trial.report` fails when the backend is `MySQL`, since `MySQL` cannot store `nan`.

## Description of the changes
1. This PR changes the `nan` to `None` to store in the backend.
2. Add tests for `nan` in `report`.

### CODE REVIEW

File: optuna/storages/_rdb/storage.py
Comment (nzw0301) [inline optuna/storages/_rdb/storage.py:56]: This can be simplified:
```suggestion
    if isinstance(value, (tuple, list)):
```

**Comment by belldandyxtq** [comment]
Should we change `None` back to `nan` in `get`?

**Comment by hvy** [comment]
Thanks for the PR.
@himkt , let me know if you're busy, but could you otherwise review this change?

**Comment by HideakiImamura** [comment]
@belldandyxtq Please apply the `formats.sh` to pass the CI.

**Review by HideakiImamura — APPROVED** [review]
Thanks for the mob review! LGTM.

[TODO] We need to add a migration script. I must be done until the next release, but it should be ASAP.

**Comment by HideakiImamura** [comment]
I fixed the conflict. 

@himkt Please merge this PR after your approval.

**Review by himkt — APPROVED** [review]
LGTM.

Note that after this PR being merged, `trial.report(step=x, value=float("nan"))` does not raise an exception without writing an intermediate entry until a storage migration. The current behavior is that stops an optimization with raising an exception. Users who stops the optimization using `trial.report` + `value=nan` may observe this behavior change (but we think no one rely on this behavior, since pruners consider trials to be pruned when `nan` exists in intermediates).

**Pull request merged**
